app/src/main/java/com/sdp/movemeet/utility/ActivitiesUpdater.java

```diff
@@ -48,91 +58,121 @@ public static ActivitiesUpdater getInstance() {
+                int size = queryDocumentSnapshots.getDocuments().size() - activities.size();
+                //add to the list only those newly added activities
+                Log.d(TAG, "activities size is " + activities.size());
+                Log.d(TAG, "diff number of activities is " + size);
+                if (size > 0) addListeners(collection.orderBy("createdAt", Query.Direction.DESCENDING).limit(size), true);
+            }
+        });
+    }
+
+    /*public void checkInternetAndFetchActivities() {
+        if (internetIsConnected()) {
+            fetchListActivities();
+        } /* else {
+            activities = cache.loadActivitiesFromCache();
+        }
+    }
+
+
+    private boolean internetIsConnected() {
+        try {
+            String command = "ping -c 1 google.com";
+            return (Runtime.getRuntime().exec(command).waitFor() == 0);
+        } catch (Exception e) {
+            return false;
+        }
+    } */
+
+    private Activity convertDocSnapToActivity(DocumentSnapshot docSnap) {
+
+        String activityId = convertObjToString(docSnap.get("activityId"));
+        String organizerId = convertObjToString(docSnap.get("organizerId"));
+        String title = convertObjToString(docSnap.get("title"));
+        int numberParticipant = convertLongToInt(docSnap.getLong("numberParticipant"));
+        double longitude = convertObjToDouble(docSnap.get("longitude"));
+        double latitude = convertObjToDouble(docSnap.get("latitude"));
+        String description = convertObjToString(docSnap.get("description"));
+        Date date = getDateObj(docSnap.getDate("date"));
+        double duration = convertObjToDouble(docSnap.get("duration"));
+        String address = convertObjToString(docSnap.get("address"));
+        Date createdAt = getDateObj(docSnap.getDate("createdAt"));
+
+        Object participantsIdobj = docSnap.get("participantsId");
+        ArrayList<String> partcipantsId;
+        if (participantsIdobj == null) partcipantsId = new ArrayList<>();
+        else partcipantsId = (ArrayList<String>) participantsIdobj;
+
+        String sportobj =  docSnap.getString("sport");
+        Sport sport;
+        if (sportobj == null) sport = Running;
+        else sport = Sport.valueOf(sportobj);; //enum sotred in firebase ?");
+
+        Activity act = new Activity(activityId, organizerId, title, numberParticipant, partcipantsId, longitude, latitude, description, date, duration, sport, address, createdAt);
+        return act;
+    }
+
+    private String convertObjToString(Object obj) {
+        if (obj instanceof String && obj != null) {
+            return (String) obj;
+        } else return null;
+    }
+
+    private int convertLongToInt(Long obj) {
+        if (obj == null) return 0;
+        else return obj.intValue();
+    }
+
+    private double convertObjToDouble(Object obj) {
+        if (obj == null) return 0;
+        else return (double) obj;
+    }
 
+    private Date getDateObj(Date date) {
```

Comment:
You might be able to make these conversion functions into a single one using generics

Reply:
That's right ! I didn't think about it, I will try to do that